File: 2025-02-February-GFG-POTD/February 25 - Histogram Max Rectangular Area/Solution.py
class Solution:
    @staticmethod
    def getMaxArea(arr):
        n = len(arr)
        leftSmall = [-1] * n
        rightSmall = [n] * n
        stack = []

        # Find the first smaller bar on the left for each bar
        for i in range(n):
            while stack and arr[stack[-1]] >= arr[i]:
                stack.pop()
            if stack:
                leftSmall[i] = stack[-1]
            stack.append(i)

        stack = []  # Clear the stack for reuse

        # Find the first smaller bar on the right for each bar
        for i in range(n - 1, -1, -1):
            while stack and arr[stack[-1]] >= arr[i]:
                stack.pop()
            if stack:
                rightSmall[i] = stack[-1]
            stack.append(i)

        # Calculate the maximum area
        maxArea = 0
        for i in range(n):
            width = rightSmall[i] - leftSmall[i] - 1
            area = arr[i] * width
            maxArea = max(maxArea, area)

        return maxArea


if __name__ == "__main__":
    t = int(input())
    while t > 0:
        arr = list(map(int, input().split()))
        res = Solution.getMaxArea(arr)
        print(res)
        print("~")
        t -= 1

File: 2025-02-February-GFG-POTD/February 25 - Histogram Max Rectangular Area/Solution.cpp
#include <iostream>
#include <stack>
#include <vector>

class Solution {
public:
    static int getMaxArea(std::vector<int>& arr) {
        int n = arr.size();
        std::vector<int> leftSmall(n);
        std::vector<int> rightSmall(n);
        std::stack<int> stack;

        // Find the first smaller bar on the left for each bar
        for (int i = 0; i < n; i++) {
            while (!stack.empty() && arr[stack.top()] >= arr[i]) {
                stack.pop();
            }
            leftSmall[i] = stack.empty() ? -1 : stack.top();
            stack.push(i);
        }

        stack = std::stack<int>(); // Clear the stack for reuse

        // Find the first smaller bar on the right for each bar
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.empty() && arr[stack.top()] >= arr[i]) {
                stack.pop();
            }
            rightSmall[i] = stack.empty() ? n : stack.top();
            stack.push(i);
        }

        // Calculate the maximum area
        int maxArea = 0;
        for (int i = 0; i < n; i++) {
            int width = rightSmall[i] - leftSmall[i] - 1;
            int area = arr[i] * width;
            maxArea = std::max(maxArea, area);
        }

        return maxArea;
    }
};

int main() {
    int t;
    std::cin >> t;

    while (t-- > 0) {
        int n;
        std::cin >> n;
        std::vector<int> arr(n);

        for (int i = 0; i < n; i++) {
            std::cin >> arr[i];
        }

        Solution obj;
        int res = obj.getMaxArea(arr);

        std::cout << res << std::endl;
        std::cout << "~" << std::endl;
    }

    return 0;
}
